src/Relation/Binary/Rewriting.agda

```diff
@@ -0,0 +1,153 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Concepts from rewriting theory
+-- Definitions are based on "Term Rewriting Systems" by J.W. Klop
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Relation.Binary.Rewriting where
+
+open import Agda.Builtin.Equality using (_≡_ ; refl)
+open import Data.Product using (_×_ ; ∃ ; -,_; _,_ ; proj₁ ; proj₂)
+open import Data.Empty
+open import Data.Sum as Sum using (_⊎_; inj₁; inj₂)
+open import Function using (flip; _∘′_)
+open import Induction.WellFounded
+open import Level
+open import Relation.Binary.Core
+open import Relation.Binary.Construct.Closure.Reflexive
+open import Relation.Binary.Construct.Closure.ReflexiveTransitive as RTrans
+open import Relation.Binary.Construct.Closure.Symmetric
+open import Relation.Binary.Construct.Closure.Transitive
+open import Relation.Nullary
+
+-- The following definitions are taken from Klop [5]
+module _ {a ℓ} {A : Set a} (_⟶_ : Rel A ℓ) where
+
+  private
+    _⟵_  = flip _⟶_
+    _—↠_ = Star _⟶_
+    _↔_  = Star (SymClosure _⟶_)
+
+  IsNormalForm : A → Set _
+  IsNormalForm a = ¬ ∃ λ b → (a ⟶ b)
+
+  HasNormalForm : A → Set _
+  HasNormalForm b = ∃ λ a → IsNormalForm a × (b —↠ a)
+
+  NormalForm : Set _
+  NormalForm = ∀ {a b} → IsNormalForm a → b ↔ a → b —↠ a
+
+  WeaklyNormalizing : Set _
+  WeaklyNormalizing = ∀ a → HasNormalForm a
+
+  StronglyNormalizing : Set _
+  StronglyNormalizing = WellFounded _⟵_
+
+  UniqueNormalForm : Set _
+  UniqueNormalForm = ∀ {a b} → IsNormalForm a → IsNormalForm b → a ↔ b → a ≡ b
+
+  Confluent : Set _
+  Confluent = ∀ {A B C} → A —↠ B → A —↠ C → ∃ λ D → (B —↠ D) × (C —↠ D)
+
+  WeaklyConfluent : Set _
+  WeaklyConfluent = ∀ {A B C} → A ⟶ B → A ⟶ C → ∃ λ D → (B —↠ D) × (C —↠ D)
+
+
+Deterministic : ∀ {a b ℓ₁ ℓ₂} → {A : Set a} → {B : Set b} → Rel B ℓ₁ → REL A B ℓ₂ → Set _
+Deterministic _≈_ _—→_ = ∀ {x y z} → x —→ y → x —→ z → y ≈ z
+
+module _ {a ℓ} {A : Set a} {_⟶_ : Rel A ℓ} where
+
+  private
+    _—↠_ = Star _⟶_
+    _↔_  = Star (SymClosure _⟶_)
+    _⟶₊_ = Plus _⟶_
+
+  det⟶conf : Deterministic _≡_ _⟶_ → Confluent _⟶_
```

Comment:
Could we use the `⇒` arrow in these proof names rather than `⟶`? As:

1. Implication is much more commonly represented using `⇒` in the rest of the library (I know there's some hold outs in `Relation.Binary.Consequences` but they'll go at some point).

2. In this particular file it's actually actively confusing as we're already using a relation `⟶` that is definitely not implication.

Reply:
That makes sense!